### PR TITLE
feature/lerp - add linear interpolation

### PR DESCRIPTION
**Pull request by mDantas**
WORK IN PROGRESS

This PR implements #549 

To do:
-add tests
-add pairwise linear interpolation
-code organization re re-used code from lib/function/geometry/distance?

### CODE REVIEW

File: lib/function/geometry/linearInterpolation.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+function _objectToArray(o){
+  var keys = Object.keys(o);
+  var a = [];
+  for (var i = 0; i < keys.length; i++) {
+    a.push(o[keys[i]]);
+  }
+  return a;
+}
```
Comment (mDantas) [inline lib/function/geometry/linearInterpolation.js]: semicolon

File: lib/function/geometry/linearInterpolation.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  return linearInterpolation;
+};
+
+function _2d(a){
+  // checks if the number of arguments are correct in count and are valid (should be numbers)
+  if (a.constructor !== Array){
+    a = _objectToArray(a);
+  }
+  return typeof a[0] === 'number' && typeof a[1] === 'number';
+}
```
Comment (mDantas) [inline lib/function/geometry/linearInterpolation.js]: semicolon

**Comment by josdejong** [comment]
Cool, thanks for picking this up @mDantas. Looks good so far

Docs in a code comment on top of the function will be welcome too :)

**Comment by mDantas** [comment]
Slllooowww and steaaaddddyy :) Sorry for the silence @josdejong - it's happening, life's just been busy. I've got a few TODOs to fill out in the tests. Comments totally welcome.

**Comment by josdejong** [comment]
Thanks Matthew, I will review your code within a few days

**Comment by SeregPie** [comment]
Any update on this. A few days have been already passed. =)

**Comment by josdejong** [comment]
whooops :astonished:

thanks for the reminder Sergej, I've totally forgotten about this PR

**Comment by josdejong** [comment]
I will review the code within a few days :wink:

**Comment by josdejong** [comment]
I think `linearInterpolation` will be a nice addition to mathjs, but this PR still needs some work. 

@mDantas can you let us know if you'r still interested in working out this PR after all this time or not?

Here some feedback points:

1. Right now the function only supports numbers. It will really start shining in math.js if it also supports BigNumbers and Fractions (and maybe even units). That shouldn't be too hard: instead of using JavaScript operators on the numeric input, you have to use mathjs functions like `math.add`, `math.subtract`, `math.divide`.

2. I miss a unit test which checks whether matrix input works, like
    ```js
    assert.deepEqual(math.linearInterpolation(1, math.matrix([-1,5]), math.matrix([4,-2])), 2.2)
    ```

3. I miss a unit test testing matrix input for the `x` value, like 
    ```js
    math.linearInterpolation([1, 2, 3], [-10, -10], [10, 10])
    ```
    and the same for matrix input like:

    ```js
    math.linearInterpolation(math.matrix([1, 2, 3]), [-10, -10], [10, 10])
    ```

4. The function `_objectToArray` is a bit tricky, it puts all keys of an object into an array. But since object keys are unordered this can give issues not ordering x and y correctly. I think we should change it to something like
    ```js
    function _objectToArray(o){
      return [o.x, o.y];
    }
    ```
5. I think we should change the check `if (a.constructor !== Array)` to `if (Array.isArray(a))`.

**Comment by gwhitney** [comment]
As this PR is years stale and never received the attention required to bring it to consistency with mathjs conventions, it is being closed. However, a complete, updated PR for a general (linear) interpolation function would certainly be reviewed for inclusion in mathjs.

**Pull request closed without merging**